herd/archExtra_herd.ml

```diff
@@ -484,7 +498,13 @@ module Make(C:Config) (I:I) : S with module I = I
 
       let do_dump_state tr st =
         pp_nice_state st " "
-          (fun l v -> do_dump_location tr l ^ "=" ^ I.V.pp C.hexa (get_val l v) ^";")
+          (fun l v ->
+           let locdump = do_dump_location tr l in
+           let valdump = I.V.pp C.hexa (get_val l v) in
+           if Misc.check_atag locdump then
+             dump_tagloc l ^ "=" ^ (get_atag_locname locdump) ^ valdump ^";"
```

Comment:
I was wondering whether it would be easy to print
memtag(y)=blue
rather than
memtag(y)=y:blue

I wonder what @maranget  thinks, but I would expect that y:blue is a tagged pointer and blue is a tag.

Reply:
Well, memtag(y) being a tagged pointer and not a tag is bizarre, but the current implementation is logical as the LDG instructions stores a tagged pointer in its target register. Consider the following 2+2W test on tagged pointers:
```
% diyone7 -arch AArch64 -variant memtag PodWW T Coe T PodWW  T Coe T -norm
% cat 2+2W+potts.litmus
AArch64 2+2W+potts
"PodWWTT CoeTT PodWWTT CoeTT"
Generator=diyone7 (version 7.56+02~dev)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Co Co
Orig=PodWWTT CoeTT PodWWTT CoeTT
{
0:X0=x:blue; 0:X1=x:red; 0:X2=y:red; 0:X3=y:green;
1:X0=y:blue; 1:X1=y:red; 1:X2=x:red; 1:X3=x:green;
}
 P0          | P1          ;
 STG X0,[X1] | STG X0,[X1] ;
 STG X2,[X3] | STG X2,[X3] ;
exists (memtag(x)=x:blue /\ memtag(y)=y:blue)

% herd7 -variant memtag 2+2W+potts.litmus
Test 2+2W+potts Allowed
States 4
memtag(x)=x:blue; memtag(y)=y:blue;
memtag(x)=x:blue; memtag(y)=y:red;
memtag(x)=x:red; memtag(y)=y:blue;
memtag(x)=x:red; memtag(y)=y:red;
Ok
Witnesses
Positive: 1 Negative: 3
Condition exists (memtag(x)=x:blue /\ memtag(y)=y:blue)
Observation 2+2W+potts Sometimes 1 3
Time 2+2W+potts 0.00
Hash=4b98e723e01cb79c7562793aae987b8d
```

And now essentially the same test, with "local observers" (here `LDG` instructions):
```
 diyone7 -obs local -arch AArch64 -variant memtag PodWW T Coe T PodWW  T Coe T -norm
% cat 2+2W+potts.litmus
AArch64 2+2W+potts
"PodWWTT CoeTT PodWWTT CoeTT"
Generator=diyone7 (version 7.56+02~dev)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Co Co
Orig=PodWWTT CoeTT PodWWTT CoeTT
{
0:X0=x:blue; 0:X1=x:red; 0:X2=y:red; 0:X3=y:green; 0:X5=y;
1:X0=y:blue; 1:X1=y:red; 1:X2=x:red; 1:X3=x:green; 1:X5=x;
}
 P0          | P1          ;
 STG X0,[X1] | STG X0,[X1] ;
 STG X2,[X3] | STG X2,[X3] ;
 MOV X4,X5   | MOV X4,X5   ;
 LDG X4,[X5] | LDG X4,[X5] ;
exists (0:X4=y:blue /\ 1:X4=x:blue)

% herd7 -variant memtag 2+2W+potts.litmus
Test 2+2W+potts Allowed
States 4
0:X4=y:blue; 1:X4=x:blue;
0:X4=y:blue; 1:X4=x:red;
0:X4=y:red; 1:X4=x:blue;
0:X4=y:red; 1:X4=x:red;
Ok
Witnesses
Positive: 1 Negative: 8
Condition exists (0:X4=y:blue /\ 1:X4=x:blue)
Observation 2+2W+potts Sometimes 1 8
Time 2+2W+potts 0.05
Hash=d3f011200da107dd032ecf34564cd033
```

Hence, perhaps a better solution simply is to change the name `memtag` into `tagged`?

Reply:
Having looked better, I see one advantage of printing tags and not tagged pointers: getting rid of special cases in printing (and probably in other places too).

Indeed, we can easily get rid of the special case (`if Misc.check_atag locdump then` etc.) for printing the `tag` primitive.
1. Print `System(TAG,s)` as `"tag(s)"` (function `pp_symbol` in `lib/constant`.)
2. Get rid of special cases (there are others).

The gain in code generality favors `tag` over `memtag`.

One should also slightly change generators to output tags in place of tagged pointers (_e.g._ `tag(x)=:blue` in place of `memtag(x)=x:blue`.

Here is an example of such a test:
```
 diyone7 -arch AArch64 -variant memtag PodWW T Coe T PodWW  T Coe T -norm
% cat 2+2W+potts.litmus
AArch64 2+2W+potts
"PodWWTT CoeTT PodWWTT CoeTT"
Generator=diyone7 (version 7.56+02~dev)
Prefetch=0:x=F,0:y=W,1:y=F,1:x=W
Com=Co Co
Orig=PodWWTT CoeTT PodWWTT CoeTT
{
0:X0=x:blue; 0:X1=x:red; 0:X2=y:red; 0:X3=y:green;
1:X0=y:blue; 1:X1=y:red; 1:X2=x:red; 1:X3=x:green;
}
 P0          | P1          ;
 STG X0,[X1] | STG X0,[X1] ;
 STG X2,[X3] | STG X2,[X3] ;
exists ([tag(x)]=:blue /\ [tag(y)]=:blue)
```
Notice the square brackets `[tag(..)]` which are logical since a memory access is performed to read the tag.

And using default printing, one should have something like:
```
% herd7 -variant memtag 2+2W+potts.litmus
Test 2+2W+potts Allowed
States 4
[tag(x)]=:blue; [tag(y)]=:blue;
...
```

Reply:
Thank you both for reviewing this.

I've updated the patches, switching over to [tag(x)]=:red as you both suggested and it indeed makes the whole implementation a lot simpler :) As I don't think we are going to ever need support for comparing against a target register value returned by an LDG, this should be perfect. I wasn't previously aware that :colour was treated by the parser as a TAG value, I had previously assumed that it would create a virtual symbol with an empty name.

I should have also addressed the rest of your comments, please let me know if you spot anything else.

Reply:
> I wasn't previously aware that :colour was treated by the parser as a TAG value, I had previously assumed that it would create a virtual symbol with an empty name.

Well I was not very aware either. Then, I noticed that the tag can be initialized _e.g._ `x.tag = :blue`, now `[tag(x)] = :blue`. Hence the idea.